src/main/scala/util/RecordMap.scala

```diff
@@ -0,0 +1,34 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util
+
+import chisel3._
+import scala.collection.immutable.ListMap
+import chisel3.internal.requireIsChiselType
+import chisel3.experimental.DataMirror.internal.chiselTypeClone
+
+final class RecordMap[T <: Data](eltMap: ListMap[String, T])
+    extends Record {
```

Comment:
```suggestion
final class RecordMap[T <: Data] private (eltMap: ListMap[String, T])
    extends Record {
```
It's not strictly necessary to make the constructor private, but if we ever wanted to refactor `RecordMap` into an interface with different implementations (eg. in Scala 2.13 add one that uses `VectorMap` instead of `ListMap`), this would make that refactoring easier.

Reply:
so then you coul donly make this with the companion object, is the idea?

Reply:
Exactly. Doing it via the companion object (aka a factory method) has a huge advantage that you can change the concrete subtype without breaking source and binary compatibility. So if we later wanted to make `RecordMap` a `trait`, that would break every use of `new RecordMap`, but the factory method would be just fine.

Reply:
Also if we wanted to do anything before or after object construction, you can do that in a factory method, can't do that in a constructor. Basically the default should always be private constructors.